calyx-py/calyx/queues.py

```diff
@@ -149,37 +142,311 @@ def __len__(self) -> int:
         return self.pifo_len
 
 
-def operate_queue(commands, values, queue, max_cmds, keepgoing=False):
-    """Given the two lists, one of commands and one of values.
-    Feed these into our queue, and return the answer memory.
+@dataclass
+class Pieo:
+    """A PIEO data structure.
+    Supports the operations `push`, `pop`, and `peek`.
+
+    Stores elements ordered increasingly by a totally ordered `rank` attribute (for
+    simplicitly, our implementation is just using integers).
+
+    At initialization we take in a set of `(int, int, int)` triples `data` which stores
+    values, ready times, and their ranks, and is ordered by rank.
+    
+    We also take at initialization a `max_len` value to store the maximum possible
+    length of a queue.
+
+    When asked to push:
+    - If the PIEO is at length `max_len`, we raise an error.
+    - Otherwise, we insert the element into the PIEO such that the rank order stays increasing.
+
+    When asked to pop:
+    - If the length of `data` is 0, we raise an error .
+
+    - We can either pop based on value or based on eligibility.
+    - This implementation supports the most common eligibility predicate - whether an element is 'ripe'.
+
+    - If a value is passed in, we pop the first (lowest-rank) instance of that value.
+    - If no value is passed in but a time is,
+        we pop the first (lowest-rank) value that passes the predicate.
+    - Note that either a value or a bound must be passed in - both cannot be, nor can neither.
+
+    When asked to peek:
+    We do the same thing as `pop`, except:
+    - We peek into the PIEO instead of popping it - i.e. we don't remove any elements.
+
+    We compactly represent these similar operations through `query`, which takes in an additional
+    optional `remove` parameter (defaulted to False) to determine whether to pop or peek.
+    """
+    
+    def __init__(self, max_len: int):
+        """Initialize structure."""
+        self.max_len = max_len
+        self.data = []
+        self.insertion_count = 0
+
+    def __repr__(self):
+        return str(self.data)
+
+    def ripe(self, val, time):
+        """Check that a value is 'ripe' – i.e. its ready time has passed"""
+        return val[1] <= time
+    
+    def binsert(self, val, time, rank, l, r):
+        """Inserts element into list such that rank ordering is preserved
+        Uses variant of binary search algorithm
+        """
+        if l == r:
+            return self.data.insert(l, (val, time, rank))
+
+        mid = (l + r) // 2
+
+        if rank == self.data[mid][2]:
+            return self.data.insert(mid, (val, time, rank))
+
+        if rank > self.data[mid][2]:
+            return self.binsert(val, time, rank, mid+1, r)
+
+        if rank < self.data[mid][2]:
+            return self.binsert(val, time, rank, l, mid)
```

Comment:
Super! And now that you have a working prototype, would you like to consider a swap to `heapq` or some other standard library?

Reply:
In light of our discussion this morning: please leave this implementation as-is, as the “verbose Python” version. You could then make a separate, super simple, library-sourced version that also agrees with this one.

Reply:
I've done this now – `NWCSimple` serves as a simple, general-purpose queue oracle for non-work-conserving queues, built as a binary heap with a few tweaks.